### PR TITLE
Update self-feeding chatbot compatibility

### PR DESCRIPTION
**Pull request by bhancock8**
Add a link to the project page and press for self-feeding chatbot project.

### CODE REVIEW

**Comment by bhancock8** [comment]
Makes sense. PEM is on leave this week, so I won't have access to a pre-trained model to give demo commands for until next week. But I may be able to at least put in stubs before then.

**Comment by jaseweston** [comment]
anu updatos?

**Comment by bhancock8** [comment]
Sorry for the delay! Running a big workshop on Mon/Tues so crunching for that. Plan to get back on this next week.

The submodule names have also changed between since the best pretrained model was trained, so it was having issues loading @pemazare reported. I think @samhumeau is taking a look at it for us to see if we can get that up?

**Comment by jaseweston** [comment]
thankooo!

**Comment by bhancock8** [comment]
Most of a README is now in place (still want to add some detail about how to chat with a pretrained model and how to do MTL training). 

Enough has changed in the parent classes of SelfFeedingAgent that it was no longer compatible with current master. I made those updates and it now runs, but I haven't had the chance yet to do full runs and confirm that the learning curves look right. Will make those updates before pinging for review.

**Comment by bhancock8** [comment]
Did some refactoring for clarity/consistency with the paper:
- replace `metadialog` and `dialogue_sf` references with `self_feeding`
- replace `explanation` references with `feedback`
- replace `sentiment` references with `satisfaction`

Also confirmed that the run commands in the README result in numbers that match the tables in the paper.

Remaining TODOs:
- update the data file (I'll coordinate with @pemazare directly on this)
- add the convai2 candidates in the proper form to that data file so interactive mode works
- upload best trained model (currently training) to a place where it can be downloaded from 
- pass linting checks

**Comment by stephenroller** [comment]
- For linting, you should just run black on everything now. That will also make the merge *much* easier.
- Add the copyright message to all the files -- this is why unit tests don't pass presently (I should move that to lint...)
- Is it reasonable to make a [gpu test](https://github.com/facebookresearch/ParlAI/blob/master/tests/nightly/gpu/test_controllable.py)? These run nightly and their goal is to verify critical numbers in the paper. It makes it loads easier for us to support your work for a long time if we know when we broke it :) Most of the GPU tests we have now just verify pretrained models.

**Comment by stephenroller** [comment]
Do you need some help pushing this over the edge to landing?

**Comment by bhancock8** [comment]
@stephenroller I certainly wouldn't say no to any help that's available! The main pieces left in terms of functionality are (a) posting a best trained model for interactive play and (b) posting the command to launch interactive mode (have an old script that probably just needs to be updated). I've got those on my list to do Wed or Thurs evening, schedule depending. But if anyone else was able to run black, add copyright messages, and/or set up a gpu test based on the full command already in the README, any of that would be helpful!

**Comment by stephenroller** [comment]
Cool, then you should focus on the changes you listed already and I'll try to help with some of those smaller merge issues in the mean time.

**Comment by stephenroller** [comment]
You'll want to skim through the change again to ensure that I didn't mess up the merge (should should've just yielded to your patch in all cases).

**Comment by bhancock8** [comment]
Thanks a bunch, @stephenroller! I got interactive mode working again and finished the README. Unfortunately, I was fighting connectivity issues tonight and couldn't finish prepping the pretrained model for download. But this PR is at a decent checkpoint and we're passing all tests. I'd vote we land this and then I'll submit another small PR to update the README once there's a model path to point people to.

**Comment by stephenroller** [comment]
That's fine with me.

File: projects/self_feeding/modules.py [hunk 1]
```diff
@@ -164,17 +164,7 @@ def score_similarity(self, context_h, cand_h):
...
-    def normalize_scores(self, scores):
-        if self.opt['scores_norm'] == 'dot':
-            return scores
-        elif self.opt['scores_norm'] == 'sqrt':
-            return scores / math.sqrt(self.opt['embedding_size'])
-        elif self.opt['scores_norm'] == 'dim':
-            return scores / self.opt['embedding_size']
-        else:
-            raise ValueError
+        return scores
```
Comment (stephenroller) [inline projects/self_feeding/modules.py]: @emilydinan can you confirm this please?

File: projects/self_feeding/self_feeding_agent.py [hunk 2]
```diff
@@ -302,65 +302,83 @@ def act(self):
...
         # If their response is a response to a rating request, no work required
         if self.status == RATING_REQUESTED:
             self.last_rating = observation['text']
         else:
             if 'text' in observation:
-                self.history.append(observation['text'])
+                self.history.update_history(observation)
 
-        if len(self.history) > 0:
+        if len(self.history.history_strings) > 0:
```
Comment (stephenroller) [inline projects/self_feeding/self_feeding_agent.py]: @emilydinan here too please

File: projects/self_feeding/self_feeding_agent.py [hunk 3]
```diff
@@ -302,65 +302,83 @@ def act(self):
...
+            assert len(set(subtasks)) == 1
             self.subtask = subtasks[0]
         # print(f"[ context ]: {observations[0]['text']}")
         return batch
 
+    def _set_text_vec(self, obs, history, truncate):
+        """
+        Set the 'text_vec' field in the observation.
+        """
+        if 'text' not in obs:
```
Comment (stephenroller) [inline projects/self_feeding/self_feeding_agent.py]: @bhancock8 this looks like a near verbatim copy of the TorchAgent `_set_text_vec`. Why is this necessary now?

File: projects/self_feeding/self_feeding_agent.py [hunk 4]
```diff
@@ -302,65 +302,83 @@ def act(self):
...
         return action
 
     def observe(self, observation):
         """Add to history, concatenate history-size utterances, and add person tokens"""
+
         # If their response is a response to a rating request, no work required
         if self.status == RATING_REQUESTED:
             self.last_rating = observation['text']
         else:
             if 'text' in observation:
```
Comment (emilydinan) [inline projects/self_feeding/self_feeding_agent.py]: you don't need this check here, calling `update_history` will check `if 'text' in observation`

File: projects/self_feeding/self_feeding_agent.py [hunk 5]
```diff
@@ -302,65 +302,83 @@ def act(self):
...
-                self.history.append(observation['text'])
+                self.history.update_history(observation)
 
-        if len(self.history) > 0:
+        if len(self.history.history_strings) > 0:
+            # WARNING: Be suspicious of this; is history_size being applied correctly?
             observation['text'] = add_person_tokens(
```
Comment (emilydinan) [inline projects/self_feeding/self_feeding_agent.py]: hmm. i don't think this is necessary. in any case we got rid of the function `add_person_tokens` in torch agent. instead, you should set the default arg `--add-person-tokens True`. then, the history will automatically be updated with this. is there a reason you did it this way?

File: projects/self_feeding/self_feeding_agent.py [hunk 6]
```diff
@@ -302,65 +302,83 @@ def act(self):
...
-                self.history[-self.opt['history_size'] :], last_speaker=1
+                self.history.history_strings[-self.opt['history_size'] :],
+                last_speaker=1,
             )
 
         self.observation = observation
 
         if observation.get('episode_done', True):
-            self.history.clear()
+            self.history.reset()
```
Comment (emilydinan) [inline projects/self_feeding/self_feeding_agent.py]: shouldn't need this here

File: projects/self_feeding/modules.py (hunk 1 above)
Comment (emilydinan) [reply projects/self_feeding/modules.py]: this looks ok to me. we typically do this outside of the module so that we can pass pre-encoded candidates to the model.

**Comment by bhancock8** [comment]
Thanks for the comments @stephenroller and @emilydinan. These have all been addressed (and marked as resolved) except the two above that pertain to methods that look similar to their counterparts in `SelfFeedingAgent`'s parent classes. The reason why `SelfFeedingAgent` overwrites the default behavior in a few places where the history is being dealt with is because it needs just a bit finer-grained control over the history (e.g., selectively slicing out turns from the history so that the context the models sees avoids turns where it asked for feedback or ratings). 

I was able to get the demo model trained, uploaded on dropbox (https://www.dropbox.com/s/8mnsljhmruj8ntq/hh131k_hb60k_fb60k_st1k.tar.gz), and referenced in the README. I presume we'll want to upload that to the same place where the other ParlAI saved models are stored and update the link. Other than that I'm not aware of any blockers to merging this in.

**Comment by stephenroller** [comment]
I'll upload the models later today and wrap up this PR (esp add a GPU test).

@bhancock8  can you share with me the metrics expected to be produced by your pretrained model?

**Comment by bhancock8** [comment]
Yep! After defining $PARLAIHOME and $MODEL (the path to the saved model), this command:
```
python $PARLAIHOME/examples/eval_model.py -mf $MODEL -t self_feeding:all:test --datatype test -bs 20
```
evaluates in ~12s and consistently returns these scores (along with others; but these are the most important ones):
```
{'dia_acc': 0.506, 'fee_acc': 0.744, 'sat_f1': 0.8343}
```

File: projects/self_feeding/README.md [hunk 7]
```diff
@@ -1,6 +1,112 @@
...
+```
+
+### Train on DIALOGUE (HH) + DIALOGUE (HB) + FEEDBACK (FB) + SATISFACTION (ST) examples
+You can train on all three tasks at once with the command below.
+```
+python -u $PARLAIHOME/examples/train_model.py -t self_feeding:all:train --model-file $MODEL -tblog true -tbcomment $MODEL -tbmetrics lr,dia_acc,dia_loss,dia_rank,fee_acc,fee_loss,sat_acc,sat_f1,sat_loss -ltim 5 -vtim 10 -vp 50 -m projects.self_feeding.self_feeding_agent:SelfFeedingAgent -cands batch --eval-candidates inline -histsz 2 --embedding-type fasttext_cc --embedding-size 300 --dict-maxtokens 250000 --num-epochs 500 --optimizer adamax --embeddings-scale false -bs 128 --relu-dropout 0 --attention-dropout 0 --n-heads 2 --n-layers 2 -lr 0.0025 --ffn-size 32 --lr-scheduler invsqrt --warmup-updates 500 --dia-train train_hh131k_hb60k.txt -vmt dia_acc -vmm max
```
Comment (stephenroller) [inline projects/self_feeding/README.md]: this failed for me. :(

```python-traceback
Traceback (most recent call last):
  File "examples/train_model.py", line 16, in <module>
    TrainLoop(opt).train()
  File "/private/home/roller/working/parlai/parlai/scripts/train_model.py", line 361, in __init__
    build_dict(opt, skip_if_built=True)
  File "/private/home/roller/working/parlai/parlai/scripts/build_dict.py", line 118, in build_dict
    world_dict = create_task(ordered_opt, dictionary)
  File "/private/home/roller/working/parlai/parlai/core/worlds.py", line 1156, in create_task
    world = create_task_world(opt, user_agents, default_world=default_world)
  File "/private/home/roller/working/parlai/parlai/core/worlds.py", line 1113, in create_task_world
    opt, user_agents, default_world=default_world
  File "/private/home/roller/working/parlai/parlai/core/worlds.py", line 1068, in _get_task_world
    task_agents = _create_task_agents(opt)
  File "/private/home/roller/working/parlai/parlai/core/agents.py", line 823, in _create_task_agents
    return create_task_agent_from_taskname(opt)
  File "/private/home/roller/working/parlai/parlai/core/agents.py", line 776, in create_task_agent_from_taskname
    task_agents = teacher_class(opt)
  File "/private/home/roller/working/parlai/parlai/tasks/self_feeding/agents.py", line 355, in __init__
    super().__init__(opt, shared)
  File "/private/home/roller/working/parlai/parlai/tasks/self_feeding/agents.py", line 218, in __init__
    super().__init__(opt, shared)
  File "/private/home/roller/working/parlai/parlai/core/agents.py", line 230, in __init__
    self.tasks.extend(create_task_agent_from_taskname(opt_singletask))
  File "/private/home/roller/working/parlai/parlai/core/agents.py", line 776, in create_task_agent_from_taskname
    task_agents = teacher_class(opt)
  File "/private/home/roller/working/parlai/parlai/tasks/self_feeding/agents.py", line 276, in __init__
    super().__init__(opt, shared)
  File "/private/home/roller/working/parlai/parlai/tasks/self_feeding/agents.py", line 74, in __init__
    raise ValueError("Unrecognized filepath: {}".format(path))
ValueError: Unrecognized filepath: /private/home/roller/working/parlai/data/self_feeding/self_feeding_v02/train_hh131k_hb60k.txt
```
```
devfair0237 parlai patch-1 » ls /private/home/roller/working/parlai/data/self_feeding/self_feeding_v02
convai2_cands.txt  test_hh.txt  train_fb_a.txt  train_fb.txt  train_hh_hb.txt  train_st.txt  valid_hh.txt
test_fb.txt        test_st.txt  train_fb_b.txt  train_hb.txt  train_hh.txt     valid_fb.txt  valid_st.txt
```

File: projects/self_feeding/README.md [hunk 8]
```diff
@@ -1,6 +1,112 @@
...
+- Append `,fee_acc,fee_loss` to the end of the `-tbmetrics` argument to include the accuracy and loss terms for the feedback task to the Tensorboard logs.
+- Add `--fee-weight 1.0` to set a coefficient by which the loss for the `feedback` task will be multiplied (the default is 1.0 already).
+
+Putting this all together, the command to recreate the 131k HH + 60k FB result from the paper is as follows (as reported in Table 9 in the paper, this setting had the same optimal hyperparameter settings as 131k HH):
+```
+python -u $PARLAIHOME/examples/train_model.py -t self_feeding:diafee:train --model-file $MODEL -tblog true -tbcomment $MODEL -tbmetrics lr,dia_acc,dia_loss,dia_rank,fee_acc,fee_loss -ltim 5 -vtim 10 -vp 10 -m projects.self_feeding.self_feeding_agent:SelfFeedingAgent -cands batch --eval-candidates inline -histsz 2 --embedding-type fasttext_cc --embedding-size 300 --dict-maxtokens 250000 --num-epochs 100 --optimizer adamax --embeddings-scale false -bs 128 --relu-dropout 0 --attention-dropout 0 --n-heads 2 --n-layers 2 -lr 0.0025 --ffn-size 32 --lr-scheduler invsqrt --warmup-updates 500 -vmt dia_acc -vmm max
```
Comment (stephenroller) [inline projects/self_feeding/README.md]: And this failed with:

```python-traceback
[ Saving tensorboard logs here: /tmp/tensorboard ]
Traceback (most recent call last):
  File "examples/train_model.py", line 16, in <module>
    TrainLoop(opt).train()
  File "/private/home/roller/working/parlai/parlai/scripts/train_model.py", line 432, in __init__
    self.writer = TensorboardLogger(opt)
  File "/private/home/roller/working/parlai/parlai/core/logs.py", line 89, in __init__
    self.writer = SummaryWriter(logdir='{}/{}'.format(tbpath, tensorboard_tag))
  File "/private/home/roller/.conda/envs/fairseq-fp16-20190211/lib/python3.6/site-packages/tensorboardX/writer.py", line 279, in __init__
    self.file_writer = FileWriter(logdir=log_dir, **kwargs)
TypeError: type object got multiple values for keyword argument 'logdir'
```

Comment (stephenroller) [reply projects/self_feeding/README.md]: dropping tensorboard stuff makes this work so I might just update the readme for now.

File: projects/self_feeding/README.md (hunk 7 above)
Comment (bhancock8) [reply projects/self_feeding/README.md]: Discussed offline. Result: the particular derivative file used in this command will be provided in the updated dataset version (rather than requiring the user to construct it themselves by concatenating the right number of lines from the corresponding files).

File: projects/self_feeding/README.md (hunk 8 above)
Comment (bhancock8) [reply projects/self_feeding/README.md]: Likely related to #1820 and a tensorboardX version issue.

**Comment by stephenroller** [comment]
```
$ python tests/nightly/gpu/test_self_feeding.py
/private/home/roller/.conda/envs/fairseq-fp16-20190211/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/private/home/roller/.conda/envs/fairseq-fp16-20190211/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
/private/home/roller/.conda/envs/fairseq-fp16-20190211/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:8: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import imp
/private/home/roller/.conda/envs/fairseq-fp16-20190211/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
...
----------------------------------------------------------------------
Ran 3 tests in 185.542s

OK
python tests/nightly/gpu/test_self_feeding.py  217.12s user 21.55s system 127% cpu 3:06.72 total
```

File: projects/self_feeding/README.md (hunk 8 above)
Comment (stephenroller) [reply projects/self_feeding/README.md]: Dropped the tensorboard arguments from the example commands to temporarily alleviate incompatibility issues.

**Comment by stephenroller** [comment]
```
$ python tests/datatests/test_new_tasks.py
.
----------------------------------------------------------------------
Ran 1 test in 2098.369s

OK
```
